cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -5,10 +5,458 @@
+
+    /** Gets the condition of this conditional expression. */
+    Expr getCondition() { none() }
+
+    /** Gets the true branch of this conditional expression. */
+    Expr getThen() { none() }
+
+    /** Gets the false branch of this conditional expression. */
+    Expr getElse() { none() }
+  }
+
+  private import semmle.code.cpp.dataflow.new.DataFlow::DataFlow as DataFlow
+  private import semmle.code.cpp.ir.dataflow.internal.DataFlowPrivate as Private
+
+  class Parameter = Cpp::Parameter;
+
+  /**
+   * A (direct) parameter position. The value `-1` represents the position of
+   * the implicit `this` parameter.
+   */
+  private int parameterPosition() { result in [-1, any(Cpp::Parameter p).getIndex()] }
+
+  /** A parameter position represented by an integer. */
+  class ParameterPosition extends int {
+    ParameterPosition() { this = parameterPosition() }
+  }
+
+  /** An argument position represented by an integer. */
+  class ArgumentPosition extends int {
+    ArgumentPosition() { this = parameterPosition() }
+  }
+
+  /** Holds if arguments at position `apos` match parameters at position `ppos`. */
+  overlay[caller?]
+  pragma[inline]
+  predicate parameterMatch(ParameterPosition ppos, ArgumentPosition apos) { ppos = apos }
+
+  final private class FinalMethod = Cpp::Function;
+
+  /**
+   * A non-overridable function.
+   *
+   * This function is non-overrideable either because it is not a member function, or
+   * because it is a final member function.
+   */
+  class NonOverridableMethod extends FinalMethod {
+    NonOverridableMethod() {
+      not this instanceof Cpp::MemberFunction
+      or
+      exists(Cpp::MemberFunction mf | this = mf |
+        not mf.isVirtual()
+        or
+        mf.isFinal()
+      )
+    }
+
+    /** Gets the `Parameter` at `pos` of this function, if any. */
+    Parameter getParameter(ParameterPosition ppos) { super.getParameter(ppos) = result }
+
+    /** Gets an expression returned from this function. */
+    GuardsInput::Expr getAReturnExpr() {
+      exists(ReturnValueInstruction ret |
+        ret.getEnclosingFunction() = this and
+        result = ret.getReturnValue()
+      )
+    }
+  }
+
+  private predicate nonOverridableMethodCall(CallInstruction call, NonOverridableMethod m) {
+    call.getStaticCallTarget() = m
+  }
+
+  /**
+   * A call to a `NonOverridableMethod`.
+   */
+  class NonOverridableMethodCall extends GuardsInput::Expr instanceof CallInstruction {
+    NonOverridableMethodCall() { nonOverridableMethodCall(this, _) }
+
+    /** Gets the function that is called. */
+    NonOverridableMethod getMethod() { nonOverridableMethodCall(this, result) }
+
+    /** Gets the argument at `apos`, if any. */
+    GuardsInput::Expr getArgument(ArgumentPosition apos) { result = super.getArgument(apos) }
+  }
+}
+
+private module GuardsImpl = SharedGuards::Make<Cpp::Location, IRCfg, GuardsInput>;
+
+private module LogicInput_v1 implements GuardsImpl::LogicInputSig {
```

Comment:
If we only have `_v1` then the suffix seems somewhat superfluous, but I assume you perhaps intend to add a `_v2`, which adds the output from Range Analysis? I.e. let's keep the `_v1` suffix.

Reply:
Yeah, eventually I will add a `_2` 😄 However, I wanted to keep things as semantically identical as I could for this first PR.